File: Difficulty: Medium/Flattening a Linked List/flattening-a-linked-list.js
//{ Driver Code Starts
"use strict";

process.stdin.resume();
process.stdin.setEncoding("utf-8");

let inputString = "";
let currentLine = 0;

process.stdin.on("data", (inputStdin) => { inputString += inputStdin; });

process.stdin.on("end", (_) => {
    inputString =
        inputString.trim().split("\n").map((string) => { return string.trim(); });

    main();
});

function readLine() { return inputString[currentLine++]; }

class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
        this.bottom = null;
    }
}

function printList(head) {
    let result = [];
    while (head !== null) {
        result.push(head.data);
        head = head.bottom;
    }
    console.log(result.join(" "));
}

function createLinkedList(arr) {
    let dummy = new Node(0);
    let temp = dummy;
    for (let i = 0; i < arr.length; i++) {
        temp.next = arr[i];
        temp = temp.next;
    }
    return dummy.next;
}

function main() {
    let t = parseInt(readLine());
    for (let test = 0; test < t; test++) {
        const n = parseInt(readLine());
        const nodes = [];

        for (let i = 0; i < n; i++) {
            const values = readLine().split(" ").map(Number);
            let dummy = new Node(0);
            let temp = dummy;

            for (let val of values) {
                temp.bottom = new Node(val);
                temp = temp.bottom;
            }

            nodes.push(dummy.bottom);
        }

        const list = createLinkedList(nodes);
        const solution = new Solution();
        const flattenedList = solution.flatten(list);
        printList(flattenedList);
        console.log("~");
    }
}
// } Driver Code Ends


// User function Template for javascript

/*LINKED LIST NODE
class Node {
  constructor(x){
    this.data = x;
    this.next = null;
    this.bottom = null;
  }
}
*/

/**
 * @param {Node} head
 * @return {Node}
 */

class Solution {
    
    flatten(head) {
        // Helper function to merge two sorted lists
        function mergeTwoLists(h1, h2) {
            // Create a dummy node to simplify list merging
            let headDummy = new Node(-1);
            let dummy = headDummy;

            // Merge the two lists while both are non-empty
            while (h1 !== null && h2 !== null) {
                if (h1.data < h2.data) {
                    dummy.bottom = h1;  // Link the smaller node to bottom
                    h1 = h1.bottom;  // Move h1 pointer down
                } else {
                    dummy.bottom = h2;  // Link the smaller node to bottom
                    h2 = h2.bottom;  // Move h2 pointer down
                }
                dummy = dummy.bottom;  // Move dummy pointer to the newly added node
            }

            // If there are remaining nodes in h1, link them
            if (h1) dummy.bottom = h1;

            // If there are remaining nodes in h2, link them
            if (h2) dummy.bottom = h2;

            return headDummy.bottom;  // Return the merged list starting from the first node
        }

        // Base case: if the list is empty or contains a single list, return the head
        if (head == null || head.next == null) return head;

        // Flatten the list starting from the next node
        head.next = this.flatten(head.next);

        // Merge the current list with the flattened list
        head = mergeTwoLists(head, head.next);

        // Set next to null as we are merging the bottom linked lists
        head.next = null;

        return head;
    }
}

File: Difficulty: Medium/Flattening a Linked List/flattening-a-linked-list.cpp
//{ Driver Code Starts
//Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

struct Node {
    int data;
    struct Node * next;
    struct Node * bottom;

    Node(int x) {
        data = x;
        next = NULL;
        bottom = NULL;
    }

};

void printList(Node *Node)
{
    while (Node != NULL)
    {
        printf("%d ", Node->data);
        Node = Node->bottom;
    }
}


// } Driver Code Ends
/* Node structure  used in the program

struct Node{
	int data;
	struct Node * next;
	struct Node * bottom;
	
	Node(int x){
	    data = x;
	    next = NULL;
	    bottom = NULL;
	}
	
};
*/

class Solution {
public:

    // Merges two sorted linked lists using the bottom pointers
    Node* merge(Node* list1, Node* list2) {
        Node* dummy = new Node(-1); // Dummy node to simplify list merging
        Node* mergedTail = dummy;   // Pointer to track the merged list

        Node* ptr1 = list1;
        Node* ptr2 = list2;

        // Merge both lists in sorted order
        while (ptr1 && ptr2) {
            if (ptr1->data < ptr2->data) {
                mergedTail->bottom = new Node(ptr1->data);
                ptr1 = ptr1->bottom;
            } else {
                mergedTail->bottom = new Node(ptr2->data);
                ptr2 = ptr2->bottom;
            }
            mergedTail = mergedTail->bottom;
        }

        // Append any remaining nodes from list1
        while (ptr1) {
            mergedTail->bottom = new Node(ptr1->data);
            ptr1 = ptr1->bottom;
            mergedTail = mergedTail->bottom;
        }

        // Append any remaining nodes from list2
        while (ptr2) {
            mergedTail->bottom = new Node(ptr2->data);
            ptr2 = ptr2->bottom;
            mergedTail = mergedTail->bottom;
        }

        return dummy->bottom; // Return the merged list, skipping the dummy node
    }

    // Flattens the linked list by merging all sublists
    Node *flatten(Node *head) {
        if (!head || !head->next) return head; // Base case: single list remains

        // Recursively flatten the rest of the lists
        Node* nextFlattened = flatten(head->next);

        // Merge current list with the flattened remainder
        return merge(head, nextFlattened);
    }
};



//{ Driver Code Starts.

int main(void) {

    int t;
    cin >> t;
    while (t--) {
        int n, m, flag = 1, flag1 = 1;
        struct Node * temp = NULL;
        struct Node * head = NULL;
        struct Node * pre = NULL;
        struct Node * tempB = NULL;
        struct Node * preB = NULL;
        cin >> n;
        int work[n];
        for (int i = 0; i < n; i++)
            cin >> work[i];
        for (int i = 0; i < n; i++) {
            m = work[i];
            --m;
            int data;
            scanf("%d", &data);
            temp = new Node(data);
            temp->next = NULL;
            temp->bottom = NULL;

            if (flag) {
                head = temp;
                pre = temp;
                flag = 0;
                flag1 = 1;
            }
            else {
                pre->next = temp;
                pre = temp;
                flag1 = 1;
            }
            for (int j = 0; j < m; j++) {

                int temp_data;
                scanf("%d", &temp_data);
                tempB = new Node(temp_data);

                if (flag1) {
                    temp->bottom = tempB;
                    preB = tempB;
                    flag1 = 0;
                }
                else {
                    preB->bottom = tempB;
                    preB = tempB;
                }
            }
        }
        Node *fun = head;
        Node *fun2 = head;

        Solution ob;
        Node* root = ob.flatten(head);
        printList(root);
        cout << endl;

    
cout << "~" << "\n";
}
    return 0;
}

// } Driver Code Ends